### PR TITLE
[Fixes #6565] Implement Luhn Loki processor

### LINKED ISSUE #6565: Add a Luhn processor
### Request

Logs that run through some PCI environments might include some credit card information that we would like to exclude from the pipeline. There are other more bureaucratic ways to solve this problem, but it would be sufficient to add a small Luhn processor to the pipeline. Since Luhn requires a specific checksum calculation for validating a variety of identification numbers it cannot be accomplished with simple regex without significant false-positives. 

A simple addition of something like https://github.com/theplant/luhn would handle all you need. 

### Use case

In a PCI environment I am processing a lot of log messages for applications that I have no control over that contain PCI information. I need to drop the PCI information before I load the logs into Loki. Our current pipeline passes through fluent-bit which allows us to write lua scripts for middleware. For now we are using a lua script in that pipeline to redact those numbers.

### PR DESCRIPTION
**Pull request by blockloop**
#### PR Description

This adds a Luhn processor to the Loki pipeline to filter credit card numbers based on the [Luhn algorithm](https://en.wikipedia.org/wiki/Luhn_algorithm)

#### Which issue(s) this PR fixes

Fixes #6565

#### Notes to the Reviewer

The unit tests pass, but I'm not 100% sure how to run some integration tests. I'm sure I'm missing some other requirement as well. Any pointers welcome.

#### PR Checklist

- [ ] CHANGELOG.md updated
- [ ] Documentation added
- [x] Tests updated
- [ ] Config converters updated

### CODE REVIEW

**Comment by mattdurham** [comment]
Thanks for the PR! Some things to add would be comments to the code (link to wikipedia on luhn maybe?) and will need some documentation for the stage added to https://github.com/grafana/agent/blob/main/docs/sources/flow/reference/components/loki.process.md

**Comment by blockloop** [comment]
@mattdurham I updated with some comments and added the docs to the link. LMK if there's anything else I need to do.

File: docs/sources/flow/reference/components/loki.process.md [hunk 1]
```diff
@@ -566,6 +568,47 @@ set of extracted data, with the value of `user=foo`.
...
+
+
+The `source` field defines the source of data to search. When `source` is
+missing or empty, the stage parses the log line itself, but it can also be used
+to parse a previously extracted value.
+
+Let's see how this works on the following log line and stages.
+
+```
+time=2012-11-01T22:08:41+00:00 app=loki level=WARN duration=125 message="credit card approved 4716415709052638" extra="user=foo"
```
[the remark this answers is not in the document]
Comment (blockloop) [reply docs/sources/flow/reference/components/loki.process.md:599]: The card has to pass the Luhn filter. I generated it with https://developer.paypal.com/tools/sandbox/card-testing/

[the remark this answers is not in the document]
Comment (blockloop) [reply docs/sources/flow/reference/components/loki.process.md:599]: I'll update with one from https://developer.paypal.com/tools/sandbox/card-testing/

[the remark this answers is not in the document]
Comment (blockloop) [reply docs/sources/flow/reference/components/loki.process.md]: Done. Let me know if that works @clayton-cornell

[the remark this answers is not in the document]
Comment (clayton-cornell) [reply docs/sources/flow/reference/components/loki.process.md]: That works :-) As long as it's one that's a public test card number or something clearly not a potentially "real" card number, we are all good.

**Comment by clayton-cornell** [comment]
@mattdurham I'm happy with the doc additions and am ready to sign off. Over to you for final code approval. :-)

**Comment by blockloop** [comment]
Thanks @clayton-cornell 

@mattdurham I wasn't able to run this to do integration testing. The tests work well but I haven't witnessed it in the pipeline. If you have any integration tests that would be nice.

**Review by mattdurham — APPROVED** [review]
Thank you, looks good to me!

**Comment by mattdurham** [comment]
There are some linting errors `validateLuhnFilterConfig is unused` along with the new, can you take another look at those?

**Review by mattdurham — CHANGES_REQUESTED** [review]
There are some unused functions that need to be resolved or hooked up.

**Comment by blockloop** [comment]
@mattdurham I realized I left out some required implementation to wire up the Luhn filter that was causing the unused code errors. I think it should be okay now.

**Review by clayton-cornell — APPROVED** [review]
Docs look ok to me

**Comment by mattdurham** [comment]
Getting a few linting errors, the easiest way to check those is by running `USE_CONTAINER=1 make lint` from the repository root.

**Comment by blockloop** [comment]
> Getting a few linting errors, the easiest way to check those is by running `USE_CONTAINER=1 make lint` from the repository root.

The linting errors don't show up for me locally, but I fixed the ones I found.

**Comment by mattdurham** [comment]
Apologies, I have it set to automerge but everytime a change is pushed I have to manually kick off our drone build, outside contributions require manual kickoff. Will monitor this today to ensure it gets pushed in.

**Comment by blockloop** [comment]
No problem. Looks like it all succeeded :)

**Review by mattdurham — APPROVED** [review]
lgtm

**Pull request merged**